### PR TITLE
[Feature] table list prototype

### PR DESCRIPTION
**Pull request by rvyuha**
1. Created a template file that imitate the tables list found at: https://github.com/Big-Life-Lab/PHES-ODM/blob/main/metadata_en.md
2. Modified the github action to install rmarkdown, and dependencies for rvml.
3. Created the qmd file from the template

Below are the code file changes

- .github/actions/build-docs-composite-action/action.yml Contains the changes to setup R and install packages
- R/rvml-mod-table.R Contains code used to create subcategories as well as subtables needed to create the different lists in tables.qmd
- qmd/tables.qmd Contains the generated output of pseudo rvml(mainly using base R)
- templates/tables-template.qmd Contains the code to generate the tables.qmd utilizing mainly base R with some packages for sql table access.

### CODE REVIEW

File: templates/tables-template.qmd [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+table_of_content <-"<ul>"
+
+for (row_index in 1:nrow(val_table)) {
+  if(is.na(val_table[row_index, ]$partType)){
+    next()
+  }
+  
+  working_row <- val_table[row_index, ]
+  
+  content_string <- paste0(content_string, working_row$partLabel, " (", working_row$partID, ") </br>", working_row$partDesc, "</br></br>")
```
Comment (yulric) [inline templates/tables-template.qmd:28]: What is this for?

File: templates/tables-template.qmd [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+  table_of_content <- paste0(table_of_content, "<li><a href=\"#", working_row$partID, "\">", working_row$partID, "</a></li>")
+  
+}
+table_of_content <- paste0(table_of_content, "</ul>")
+
+
+
+quary <- paste("SELECT", "*", "FROM", "cat_tables")
+val_table <- DBI::dbGetQuery(main_db, quary)
+new_content <- ""
```
Comment (yulric) [inline templates/tables-template.qmd:39]: Rename to `tables_string`

File: templates/tables-template.qmd [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+      paste0(
+        new_content, "<ul>")
+    }
+  }else{
+    required_col_name <- paste0(running_table_name,"Required")
+    
+    tmp_key_type <- working_row[[running_table_name]]
+    if(length(tmp_key_type)>0&&!is.na(tmp_key_type)) {
+      if (tmp_key_type == "fK") {
+        tmp_key_type <- "Foreign Key"
```
Comment (yulric) [inline templates/tables-template.qmd:78]: Don't mutate, set to a new variable

**Review by yulric — CHANGES_REQUESTED** [review]
* I think there's an error with the measures link in the table of contents. It leads to an entrely new page.
* I'm not sure that the categories are being corectly displayed. Could regenerate the table with the dictionary file I attached to the task and post a screenshot of the `samshed` column in the sites table

File: R/rvml-mod-table.R [hunk 4]
```diff
@@ -1,6 +1,6 @@
-#mod_sql_table("rvml/main.sqlite", "cat_info", "parts")
+#mod_sql_table_sub_categories("data/tables/main.sqlite", "cat_info", "parts")
```
[the remark this answers is not in the document]
Comment (rvyuha) [reply R/rvml-mod-table.R:1]: This was a comment for me to quickly be able to reuse the function call during testing.

File: R/rvml-mod-table.R [hunk 5]
```diff
@@ -63,4 +63,94 @@ RSQLite::dbWriteTable(
...
+}
+
+
+# mod_sql_table_table_of_content("data/tables/main.sqlite", "parts", "table_of_contents", "cat_tables")
+mod_sql_table_table_of_content <- function(main_db_path, in_table_name, out_table_name, out_table_with_categories_name){
+  
+  main_db <- DBI::dbConnect(RSQLite::SQLite(), main_db_path)
+  quary <- paste("SELECT", "*", "FROM", in_table_name)
+  val_table <- DBI::dbGetQuery(main_db, quary)
+  val_table <- val_table[grepl("^NA", rownames(val_table))==F,]
```
[the remark this answers is not in the document]
Comment (rvyuha) [reply R/rvml-mod-table.R:75]: This used to work but unfortunately while fixing another bug it broke its functionality. I will attempt to remove or fix it.

File: templates/tables-template.qmd (hunk 1 above)
Comment (rvyuha) [reply templates/tables-template.qmd:28]: To create a bigger space after part Description as it was cluttered when it was only one line break.

Comment (rvyuha) [reply templates/tables-template.qmd:28]: To increase the space after a description for clarity a single line break was too cluttered.

**Review by rvyuha — COMMENTED** [review]
Accidentally started a review while trying to reply to comment

File: R/rvml-mod-table.R [hunk 6]
```diff
@@ -63,4 +63,94 @@ RSQLite::dbWriteTable(
 )
 
 DBI::dbDisconnect(main_db)
+}
+
+
+# mod_sql_table_table_of_content("data/tables/main.sqlite", "parts", "table_of_contents", "cat_tables")
```
[the remark this answers is not in the document]
Comment (rvyuha) [reply R/rvml-mod-table.R:69]: As mentioned above this is a saved function call to make it easier for me to debug and test.

**Comment by rvyuha** [comment]
> * I'm not sure that the categories are being corectly displayed. Could regenerate the table with the dictionary file I attached to the task and post a screenshot of the `samshed` column in the sites table

image

File: .github/actions/build-docs-composite-action/action.yml [hunk 7]
```diff
@@ -8,6 +8,13 @@ runs:
     - name: Install tinytex
       run: quarto install tool tinytex
       shell: bash
+    - name: setup-r
+      uses: r-lib/actions/setup-r@v2
+      with:
+        r-version: "4.1.1"
+    - name: Install rmarkdown package
+      run: R -e 'install.packages(c("rmarkdown", "DBI", "RSQLite"))'
```
[the remark this answers is not in the document]
Comment (rvyuha) [reply .github/actions/build-docs-composite-action/action.yml:16]: I am not sure I am qualified to write a DESCRIPTION to merge to main.

**Review by DougManuel — CHANGES_REQUESTED** [review]
1) I added an introduction. Some of this introduction is a temporary 'explanation' until we generate those separate documents. 
1) Add version of the dictionary (I added this to the beginning) and date created (at the end).
1) I have made edit changes to the description for some tables in the excel dictionary.
1) The formatting would benefit from edits. @rvyuha @yulric @mathew-thomson please provide suggestions. 

**For the table description.** 
-Add `instr`.
Currently,
```
Address Table (addresses) 
The table that contains information about addresses, which can be fore Sites, individuals reporting data, or organizations.
```

Change to, 
```
Address Table (addresses) 
The table that contains information about addresses, which can be fore Sites, individuals reporting data, or organizations. The Sites, Organizations, and Contacts tables include a link to the addresses table through Address ID.
```

**For the header information.** 
Currently, 
```
Address ID (addID):(Primary Key) (Required: mandatory) [varchar]
A unique identifier for an address.
```

I suggest

```
Address ID (addID). A unique identifier for an address. Role: Primary Key, mandatory. Data type: varchar. 
```

**Comment by DougManuel** [comment]
The parts look-up table is missing the list of headers. (This is probably a typo in the dictionary. I did find typos in a few of the corresponding part names for heads and I corrected those, but there may be more errors. Check to make sure you are using the latest version of the dictionary and send a note in the PR or Teams if you identify more errors that need updating.)

Screenshot 2023-01-25 at 7 19 57 AM

**Comment by DougManuel** [comment]
the entries '..Order' and '..Require` should be removed they are not part of the table. They are metadata to describe what parts or in the table and the order of headers.  Some tables have '..Order' and '..Require` headers  they do not contain the actual list of headers for the table. 

The  '..Order' and '..Require`descriptions should **not be included**, but the list of the headers **should be included**. 

Tables that have '..Order' and '..Require` but not the list of headers inlcude: Contact, Protocol organization, Protocol steps, Sample relationships. See below.

Screenshot 2023-01-25 at 7 21 20 AM

**Comment by DougManuel** [comment]
The most recent dictionary is in OSF.io 'working copies' component. https://osf.io/sxuaf/

**Comment by mathew-thomson** [comment]
Sorry for the late review here, team - I reviewed @DougManuel 's suggestions above and agree with all of them. Some additional notes on this:

1. First of all - I think this is looking really good! I'm actually really impressed with what I'm seeing. I think we have some organizational edits and content shifting to do, but it has good bones and I think you've done a great job. Nice work, @rvyuha !
2. @DougManuel - I know this introduction section is meant to be a placeholder, but did we want this to be a final-ish product? Like what we would want a polished, final product to look like? I am inclined to think "yes" since we'll probably need users to refer to this document sooner rather than later.
3. I think there needs to be a heading for the hyperlinked table of contents (nit-picky, I know), because right now it sort of reads as though the list of tables are all part of "5) Dictionary tables"
4. Also, we might also want this table of contents to use the labels (ie. 'Address table') rather than just the part ID, to make sure user know what they're looking at.
5. Echo Doug's point that we probably want to reformat how the headers are described. I also think it would be beneficial to specify that the code in parentheses beside a given header is a part ID for that table, just to be clear to the reader. so `Address ID (addID):` becomes `Address ID (partID: addID):` or something to that effect. Moving the `(Primary Key) (Required: mandatory) [varchar]` to the end of the description, and including instructions would be good too.
6. I noticed that the headers are listed in alphabetical order. I think this is good, but given that we have an "order" column for each table in the parts list, I wonder how hard it would be to list the headers in the order specified in those columns, rather than in alphabetical order.
7. I noticed there's a table there that I'm not sure where it came from - 'organizationType'. It should not be there. The 'reporter' table is also listed, but this table has been deprecated in version 2.0. Would it be a huge hassle to also filter on the `status` column for a value that says `active` when filtering for the tables?
8. Doug already pointed out the parts table having no headers, and some other tables having the order and required columns instead of headers - just a small bug, I'm sure.
9. Seeing some weirdness in other tables as far as headers too - if we could update the parts list used to generate the qmd and html files and see if that fixes the issues that would be helpful.

I think this is all my comments for 'tables' for now. Again, I think this is looking really good and with a few tweaks and an updated parts list we should be good to go in no time. Great work!

File: qmd/tables.qmd [hunk 8]
```diff
@@ -1,9 +1,205 @@
...
+```{r, echo=FALSE, warning=FALSE}
+# Can put warning=FALSE above if warnings in display are an issue
+# Source utility functions
+source("../R/qmd-utils.R")
+
+parts_sheet_name <- "parts"
+path_to_excel <- "../data/raw/ODM-dictionary.xlsx"
+
+# Declaring string that will be appended during the loops
+table_of_content <- glue::glue('# Table of Contents
```
Comment (yulric) [inline qmd/tables.qmd:42]: This should be using a second level heading, ## since the Tables header at the very top is a single heading

**Review by yulric — CHANGES_REQUESTED** [review]
1. I think we should break up the code into three main sections. 
    1. Import the parts sheet
    2. Format the parts sheet. Here we would do all the processing that's being done within the code like removing any rows with missing partIDs, making columns lower case etc.
    3. And finally render the tables for the user
2. Can you Roxygen comments to all the functions in the utils folder
3. Can we remove all the rvml code that was generating the tables list from the package
4. It looks like the build-docs action is failing. Can you fix it?
5. We don't need the tables-template.qmd file right?

**Review by DougManuel — CHANGES_REQUESTED** [review]
The formatting generally looks very good.
 - `Protocol relationships table(protocolRelationships)`. Add a space between 'table' and '('.
 - Can we remove the `partID` label?
 `Protocol Step ID (partID: stepID).` to `Protocol Step ID (stepID).`

I note a few text edits we'll need to make, but @mathew-thomson @il43  or I can make those changes.

We'll want to add links to the partID, sets. Can you add the link to the partID now? And we can add the links to categories and sets after we finalize those reference documents

File: qmd/tables.qmd [hunk 9]
```diff
@@ -1,9 +1,205 @@
+<!--Source constants file until I get clarity for environment usage across entire package or if sourcing is enough -->
+
+`r source("../R/package_constants.R")`
+
+<!--` reference1 -->
+
 # Tables {.unnumbered}
 
-The list of tables in the three main groups (partType = tables, class = results, prgramDesc, lookup).
+The list of tables below are for ODM version `r pkg.env$dictionary_version_used`.
```
[the remark this answers is not in the document]
Comment (DougManuel) [reply qmd/tables.qmd:9]: We don't have all our parsing completed yet. I think we'll want to revisit after the next stage where we generate our versioning and documentation workflow.

**Comment by DougManuel** [comment]
> I think we should break up the code into three main sections.
> 1. Import the parts sheet
> 2. Format the parts sheet. Here we would do all the processing that's being done within the code like removing any rows with missing partIDs, making columns lower case etc.
> 3. And finally render the tables for the user

I agree that it would be helpful to section the code. At this point, it would be hard for someone else to generate lists, but folks could edit and modify the code. 

- there are sections such as the following that I, too, have written when generating draft code for the documentation workflow. I suggest that we start to create a list of elements that are reused and put the generation of these elements in a separate, common utilities and elements. For example, we'll want a data frame for parts in the documentation repo.
```
parts_sheet_name <- "parts"
path_to_excel <- "../data/raw/ODM-dictionary.xlsx"
```

- similarly, I wonder about data cleaning functions that would be reused between reference documents and lists. Could the following be a function that is reused in all tables. 

I pose these design questions because a clear tool kit and uitilty functions will help people generate their own lists.

**Comment by DougManuel** [comment]
The text display is close to finished. My preference is to push to main in the next couple of days and then either keep the PR open or make a new pr to address the code discussion. 

I would like our EU partners to review the table documentation. They will appreciate reviewing on the websites (rather than a PR).

**Comment by yulric** [comment]
> The formatting generally looks very good.
> 
> * `Protocol relationships table(protocolRelationships)`. Add a space between 'table' and '('.
> * Can we remove the `partID` label?
>   `Protocol Step ID (partID: stepID).` to `Protocol Step ID (stepID).`
> 
> I note a few text edits we'll need to make, but @mathew-thomson @il43 or I can make those changes.
> 
> We'll want to add links to the partID, sets. Can you add the link to the partID now? And we can add the links to categories and sets after we finalize those reference documents

re your point about removing `partID` label, @mathew-thomson suggested we do that. Should we confirm some of these changes at the next documentation meeting so there's no confusion?

**Comment by DougManuel** [comment]
> re your point about removing `partID` label, @mathew-thomson suggested we do that. Should we confirm some of these changes at the next documentation meeting so there's no confusion?

Yes, let's wait to hear from @mathew-thomson

**Comment by DougManuel** [comment]
- Could you list the tables in alphabetical order? 
     - In a future update, we may want to group the tables in those four categories, but we don't have those classes identified yet in the ODM dictionary.
- Could you create persistent links for the tables. i.e. docs.phes-odm.org/tables.html#measures

**Comment by DougManuel** [comment]
> 
> I pose these design questions because a clear tool kit and uitilty functions will help people generate their own lists.

I just realized that you have the package constants file. Looks good. Beyond constants, there are assets like the parts table and utility functions that can be reused. 

I created a few functions here. https://github.com/Big-Life-Lab/PHES-ODM-Doc/blob/OSF/R/version_utils.R

And started utilities for working with fils on OSF.io here. https://github.com/Big-Life-Lab/PHES-ODM-Doc/blob/OSF/R/osf_util.R

**Comment by DougManuel** [comment]
We'll need persistent links for all the tables -- and other parts -- and making is clear that there is a persistent link is challenging. 

qmd header uses an anchor like this: 

`partID: partType.<a class="anchored" data-anchor-id="test"></a>More text here!`

But I wonder if we should make the table names an H6 header on its line. 

What do you think?

File: qmd/tables.qmd [hunk 10]
```diff
@@ -1,9 +1,205 @@
...
+# Can put warning=FALSE above if warnings in display are an issue
+# Source utility functions
+source("../R/qmd-utils.R")
+
+parts_sheet_name <- "parts"
+path_to_excel <- "../data/raw/ODM-dictionary.xlsx"
+
+# Declaring string that will be appended during the loops
+table_of_content <- glue::glue('# Table of Contents
+                               <ul>')
```
[the remark this answers is not in the document]
Comment (rvyuha) [reply qmd/tables.qmd:63]: In sets

**Comment by rvyuha** [comment]
@yulric Ready for re-review

**Comment by rvyuha** [comment]
@yulric Ready for re-review.

File: R/qmd-utils.R [hunk 11]
```diff
@@ -0,0 +1,63 @@
...
+#' @param part_ID Id of the part in string to check, optional parameter used in construction of the warning.
+#' @param string_source_name Name of the table string_to_check comes from, optional parameter used in construction of the warning.
+#' @param optional_warning A warning overwrite string to bypass warning creation when partID or table name are missing.
+#' 
+#' @return string_to_check as is or modified into "" if it fails the check.
+format_input <-
+  function(string_to_check,
+           part_ID = NULL,
+           string_source_name = NULL,
+           optional_warning = NULL) {
```
Comment (yulric) [inline R/qmd-utils.R:38]: This can just be `warning`

**Review by yulric — CHANGES_REQUESTED** [review]
1. I expected all the format_input calls to be made in the formatting section I mentioned in my previous review
2. I don't think we need the pkg.env to be a new environment, it can just be a named list. In addition I think we should rename pkg.env to constants.
3. package-constants.R can just be constants
4. Delete debug.R

**Pull request merged**